indico/web/client/js/react/components/time/time_value.js

```diff
@@ -0,0 +1,226 @@
+    const hourString = match.groups.hours;
+    const minuteString = match.groups.minutes ?? '0';
+    const isPM = !!match.groups.meridiem?.toLowerCase().includes('p');
+    const hour = meridiemFixups[isPM](parseInt(hourString, 10));
+    const minute = parseInt(minuteString, 10);
+    return [hour, minute];
+  } else {
+    return [NaN, NaN];
+  }
+}
+
+function parse24h(timeOfDayString) {
+  let match;
+
+  if ((match = T24H_PATTERN.exec(timeOfDayString))) {
+    const hourString = match.groups.hours;
+    const minuteString = match.groups.minutes ?? '0';
+    const hour = parseInt(hourString, 10);
+    const minute = parseInt(minuteString, 10);
+    return [hour, minute];
+  } else {
+    return [NaN, NaN];
+  }
+}
+
+const parsers = {
+  '24h': parse24h,
+  '12h': parse12h,
+};
+
+/**
+ * Time represents the time of day or a duration within
+ * a 24-hour period.
+ *
+ * Internally, the value is represented by a single number, number
+ * of minutes since midnight. This class is used mostly to bundle
+ * together convenience methods, and add clarity (compared to using
+ * a naked number).
+ */
+export class Time {
+  static MAX_TIME = 1439; // last minute of the day, 23:59 in minutes since 0:00
+
+  static fromHour(hour, minute) {
+    console.assert(
+      isNaN(hour) || (hour >= 0 && hour <= 23),
+      `Hour must be between 0 and 23, got ${hour}`
+    );
+    console.assert(
+      isNaN(minute) || (minute >= 0 && minute <= 59),
+      `Minute must be between 0 and 59, got ${minute}`
+    );
+    const minutesSinceMidnight = Math.min(
+      this.MAX_TIME,
+      Math.max(0, Number(hour) * 60 + Number(minute))
+    );
+    return new this(minutesSinceMidnight);
+  }
+
+  /**
+   * The input string can be in one of the following formats:
+   *
+   * - HH:MM (zero-padded hours)
+   * - H:MM (non-zero-padded hours)
+   * - HH:M (non-zero-padded minutes)
+   * - H:M (non-zero-padded hours and minutes)
+   * - HHMM (no colon)
+   * - HMM (non-zero-padded hours and no colon)
+   * - HH:MMa (zero-padded hours, 'a' is any string with an 'a' or 'p')
+   * - H:MMa (non-zero-padded hours, 'a' is any string with an 'a' or 'p')
+   * - HHa (no minutes, 'a' is any string with an 'a' or 'p')
+   * - Ha (non-zero-padded hours without minutes, 'a' is any string with an 'a' or 'p')
+   *
+   * Note that this function will also parse 'HM' (single digit hour,
+   * single digit minutes) if the combination is valid (e.g., '56').
+   *
+   * The separator can be ':', '.', or 'h'. There is no support for Korean,
+   * Burmese, and several other locales that use either non-Arabic numerals
+   * or non-Latin meridiem indicators, or locaes where meridiem indicators
+   * are prefixed.
+   *
+   * How this is interpretted depends on the format parameter which can be
+   * '24h' or '12h' (default is '24h').
+   *
+   * See tests for a more complete list of examples.
+   */
+  static fromString(timeOfDayString, format = '24h') {
```

Comment:
When I type e.g. `23:00` in a field in am/pm mode it gets rejected, instead of being turned into `11:00 PM`

Reply:
Yes, I'd prefer that. Especially when you use a system that's using am/pm as someone who's used to 24h format, it's much more pleasant to be able to type times in a decent format :)